Proxy Data Socket

kb1lqc opened the pull request:
This branch is experimental and adds a socket server to Proxy which provides TX and RX capabilities for a Faraday radio. It also handles multiple radios and will automatically assign appropriate ports albeit a bit hardcoded for now. We should look at this PR as more of a bleeding edge rather than polished product. I'll be adding documentation for it shortly in a follow up commit/PR.

## Changelog
* Adds a socketworker thread which creates a TX socket for each radio, currently hardcoded to port 10000 and will accept data into the socket which will promptly be transmitted by Faraday.
* A second socketworker RX thread starts on port + 10 to handle RX from faraday buffers which it then transmits out over the socket.
* A bufferworker thread moves data between a buffer the sockets send/receive data with and Proxy buffers getDicts and postDicts.
* bufferworker does not currently handle partial data well, this is a known bug that needs to be fixed before the PR is accepted.

Connecting to the port 10000 with a client and sending data will result in Faraday transmitting the data. Any received data on the same Faraday will show up on port 10010 and a call with recv() after sending a single byte of data (weird) will result in the received data streaming to the socket 10010. Two Faraday radios are supported and may communicate with each other (broadcast) while connected to a multi-unit configured Proxy application.

@jbemenheiser @reillyeon @el-iso

kb1lqc committed:
Initial socket Server Created

Only sends a welcome message but hey it's a start. Also commented out
Flask.

kb1lqc committed:
Basic socket reception and saving to buffer

just a simple socket that unpacks BASE64 data and appends it to a buffer.

kb1lqc committed:
Packing individual bytes into base64 and unpacking
Please enter the commit message for your changes. Lines starting

kb1lqc committed:
Commit with Brent here so we can work on this together

kb1lqc committed:
Posting to postDicts and popping off

kb1lqc committed:
Sending data over Proxy POST and RF!

LED is blinking, we have RF being sent! Had to hack in some 00 control
bytes to get the UART to work right and Faraday know to send these over RF
as data packets.

kb1lqc committed:
Removed uartworker commented out code

kb1lqc committed:
Cleaned up socket_worker

Commented code a bit and reorganized to make a bit more sense. Still
crappy code but meh, MAKING MOVES SON!

kb1lqc committed:
Cleaned up bufferWorker

Removed some commented code and an entire While loop...

kb1lqc committed:
Cleaned up bufferworker more

Undid hardcode for unit and added comments to better define code. Removed
some more unused code.

kb1lqc committed:
Working transmissions on two different units

Really hacked code from @kb1lqd and myself which creates seperate buffer
queues for each unit to transmit. This lets us use one computer to to a
TX/RX test.

kb1lqc committed:
FINALLY WORKING RECEIVE OVER DATA SOCKETS

I separated the threads so we have a different socket +10 higher than RX
socket that is transmitting the receive proxy data from the RF link.
Not sure if this was 100% necessary but I did it. Then I realized I HAD to
send a byte of data (" ") to the receive socket before I could get a
response from it with the RX data from getDicts.. Wierd.

kb1lqc committed:
Cleaned up data receiving code

Created helper functions to compartmentalize the act of receiving data.

kb1lqc committed:
Connect and reconnect to server to send!

Can succesfully use a client application to connect and reconnect to the
socket to send data to the server without restarting proxy.

kb1lqc committed:
Somewhat robust RX port (sending) code

I can now reconnect to the port to continue getting data. It also better
handles popping from an empty dequeue but I still need to do a little
empty buffer and error handling.

kb1lqc committed:
Fixed a few Proxy socket exceptions and no data cases

I attempt to respond with a "No Data!" string and close the connection
if the queue has no data in it.

kb1lqc committed:
Cleaned up comments in socket workers

Both socket workers have a few more comments to help explain operation.

kb1lqc committed:
Removed BASE64 external socket expectations

No longer is it expected to send or receive BASE64 from the socket. Just
send something.

kb1lqc committed:
Fixed reconnect issue with socket RX

Added correct length check for an empty queue and safely closed socket
followed by exiting the loop.

kb1lqc committed:
Fixed no data ever received error and cleaned up logger prints

Caught an exception for when no nata has ever been received and the
queue doesn't exist. Now you can connect a client to the RX port and be
informed if no data exists in this case too.

kb1lqc committed:
Cleaned up socket_worker thread starting code

Removed unused arguments.

kb1lqc committed:
Cleaned up logger prints

kb1lqc committed:
Updated logging info with faraday name and removed socket exc logging

Now displays more data for each port connected as well as removing a
socket.error logger exception statement which just cluttered the screen
before closing the socket.

kb1lqc committed:
Cleaned up code to pass PyTest

Code now passes our current pytest settings. Also commented about flask
server starting up and uncommented it.

kb1lqc committed:
Moved dataBuffer creation into main and passed to threads

kb1lqc committed:
Moved unitDict into main

kb1lqc committed:
Moved getDicts and postDicts back due to flask needs

Flask needs access and I don't know how to pass flask calls arguments
from main so we'll deal with these being global for now.

kb1lqc committed:
Cleaned up Pytest failures

Too many spaces in a section of proxy

kb1lqd commented:
Issue 1

If I connect a client and don't send data but close the client then proxy crashes. The client program opens a socket with proxy but waits for user input before sending. I don't send data and just close:

 
![image](https://user-images.githubusercontent.com/339144/28809185-8fdfd9a8-7636-11e7-9eb5-12b7e3b720a8.png)

Closing the client without sending causes the 'data' references before assignment crash.

![image](https://user-images.githubusercontent.com/339144/28809195-a7590c26-7636-11e7-91cc-812a138fccc4.png)

kb1lqd commented:
# Issue 2

Closing a connection after sending any data causes Proxy to continuously transmit (I assume the last item in the buffer):

Proxy displays: 

![image](https://user-images.githubusercontent.com/339144/28809225-edf13d2a-7636-11e7-86f9-9e399109ac68.png)


I must close proxy to stop transmitting.

kb1lqc committed:
Fixed while loop break in receivedata

I never exited the while loop in receive data which caused a bug in
Faraday-Proxy that when you sent data and received it but then closed the
transmitting connection the program would get stuck in an infinit loop and
also not stop transmitting.

kb1lqc commented:
I can't reproduce your error when not sending data. Pretty sure https://github.com/FaradayRF/Faraday-Software/pull/251/commits/ccedebad374ae6ea777c49af42b202c4855adb6d fixed both your stated issues.

kb1lqd commented:
Issue 1 fixed confirmed.

Issue 2 cannot reproduce either. I was due to an older installation being present. Disregard.

kb1lqd commented:
# Issue 3

This program should return the exact bytestream assumption a socket implies. The program data should include in EVERY proxy packet layer 4 a simple packet that has payload length so that the exact payload and only that payload can be returned.

Example:
Transmitting a small amount of data like "Test" should only return "Test" and not followed by the remaining padding.

![image](https://user-images.githubusercontent.com/339144/28809812-9381093e-763a-11e7-8ba4-4b3d5302e6e2.png)


Should NOT return the padding:

![image](https://user-images.githubusercontent.com/339144/28809864-ec4fa0c0-763a-11e7-8294-ba0066682053.png)

kb1lqc committed:
Basic working stream

Limited proxy posting data to chunks of 39 bytes. This is a MSP430
hardware limitiation that will exist until we write a software based
packet handler for the MSP430. This code needs some cleanup.

kb1lqc committed:
Implemented configuration file payloadsize, clean

Cleaned up some logging statements and also implemented a config INI file
use of setting payload size. I removed all traces of the payloadSize_main
global variable @kb1lqd added as a test.

kb1lqc committed:
Cleaned up per pytest

Added spaces around arithmatic per pytest output.

kb1lqd commented:
Confirmed Issue 3 resolved:

![image](https://user-images.githubusercontent.com/339144/28832500-7cb651e4-7691-11e7-855f-c58e865718e6.png)

![image](https://user-images.githubusercontent.com/339144/28832503-80a1119a-7691-11e7-9199-f69e4a45f932.png)

The receiver socket worker was updated to return only the data that was placed into the socket. Fragmentation is had for free due to the queue functionality.

kb1lqd commented:
Issue 4 (@KB1LQC)

When testing the transmission and reception I found that the receiver client when quitting occasionally caused this error in proxy that prevented re connection (receiver only) until proxy was restarted.

![image](https://user-images.githubusercontent.com/339144/28832789-48f8557c-7692-11e7-95ae-91afeee5f3b7.png)

kb1lqd commented on faraday/proxy.py:
```diff
@@ -393,6 +428,208 @@ def testdb_read_worker():
+        logger.error(e)
+
+
+def extractBytes(data, dataBuffer, unit):
```
Is this really possible? That doesn't seem right... Encoding relies on stuffing multiple bytes for single bytes as needed...

kb1lqc replied:
What?

kb1lqd commented on faraday/proxy.py:
```diff
@@ -393,6 +428,208 @@ def testdb_read_worker():
+def sendData(conn, addr, getDicts, unit, payloadSize):
+    while True:
+        # not sure why I need a delay... otherwise socket closes
+        time.sleep(0.05)
```
We should let this delay go for now but on the next sweep we should eliminate bottlenecks and hopefully polling as much as possible. At minimum "batch" the functions to execute until queue depleted with a polling function only after queue empty.

kb1lqc replied:
This was partially addressed in https://github.com/FaradayRF/Faraday-Software/pull/251/commits/90b20a67aea82a58978c63f926bf3760e7498a02 as I removed the sleeps and now the send code is blazing along. Very sure Faraday firmware is limiting here. Or at least UART code.

However, I noticed a bug where we can't connect to the RX port and disconnect without ever sending data to the TX port (so RX is populated). This is a weird bug...

kb1lqd commented on faraday/proxy.py:
```diff
@@ -393,6 +428,208 @@ def testdb_read_worker():
+
+        # unpack frames and retrieve data originally sent to socket
+        try:
+            dataList = struct.unpack("BB121s", data)
```
Can you bring this struct format to a global definition? This would better organize all relevant formatting IMHO.

kb1lqd replied:
Possible to leave this for the next PR to optimize/clean since there are others like this below.

kb1lqd commented on faraday/proxy.py:
```diff
@@ -393,6 +428,208 @@ def testdb_read_worker():
+        try:
+            dataList = struct.unpack("BB121s", data)
+            dataList2 = struct.unpack("B{0}s".format(payloadSize), dataList[2][:payloadSize + 1])
+            socketData = dataList2[1][:dataList2[0]]
```
Add comment to describe that dataList2[0] is the packed length in bytes of the payload data.

kb1lqd commented on faraday/proxy.py:
```diff
@@ -393,6 +428,208 @@ def testdb_read_worker():
+                break
+
+
+def socket_worker(modem, getDicts, dataPort, dataBuffer):
```
Like RX, maybe rename to "socket_worker_TX"?

kb1lqd commented on faraday/proxy.py:
```diff
@@ -393,6 +428,208 @@ def testdb_read_worker():
+    try:
+
+        payload = ''.join(temp)
+        preamble = struct.pack("BB", 0, 0)  # Preamble is two 0's at this time
```
Like above, these packet struct formats should probably be located at the global level as organized single locations to see both TX and RX formatters.

kb1lqd replied:
Also, may want to comment in code that these two bytes for "preamble" is not really preamble but just packet header for communicating with the MSP430 firmware. They are packet type and sequence number expected by the msp430 firmware and slated for future use.

kb1lqc replied:
Updated with better comment and a TODO statement in 59b82cf. I want to not jump to a bunch of globals at the moment.

kb1lqd commented on faraday/proxy.py:
```diff
@@ -827,11 +1068,31 @@ def main():
+
+                v.start()
+            except:
+                logger.error("crap")
```
Should probably remove this...

kb1lqc replied:
Removed in 27819ff

kb1lqc commented:
Looking into why we drop packets @kb1lqd  I noticed the following:

```
2017-08-01 22:10:27,651 - Proxy - INFO - 123
2017-08-01 22:10:27,651 - Proxy - INFO - '\x00\x00\x1d24, 25, 26, 27, 28, 29, 30, 3 16, 17, 1\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff'
2017-08-01 22:10:27,703 - Proxy - INFO - 42
2017-08-01 22:10:27,703 - Proxy - INFO - '\x00\x00#1, 32, 33, 34, 35, 36, 37, 38, 39, 7, 1'
2017-08-01 22:10:27,707 - Proxy - WARNING - unpack requires a string argument of length 123
2017-08-01 22:10:27,709 - Proxy - WARNING - 42
2017-08-01 22:10:27,709 - Proxy - WARNING - '\x00\x00#1, 32, 33, 34, 35, 36, 37, 38, 39, 7, 1'
2017-08-01 22:10:27,766 - Proxy - INFO - 123
2017-08-01 22:10:27,766 - Proxy - INFO - '\x00\x00"40, 41, 42, 43, 44, 45, 46, 47, 48\x00\xc0\x1e\x00\x00\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff\xff'
```

The above output is from always printing the length of the data to be unpacked as well as the data itself, the warning output is the length and data when struct.unpack fails.

Do you find it odd that we get length 42 data which will obviously fail a struct.unpack that is expected 123 data of which 121 bytes is the payload? Is your uart layer doing some of the fragmentation with RF? Should we be looking for length 42 and acting accordingly?

Edit:

BOOM look at that, it must be. The missing/dropped line is perfectly equal to the missing data between the RX test client received output...

![image](https://user-images.githubusercontent.com/331540/28858752-a436e442-7707-11e7-97fe-3a4b6e20111f.png)

kb1lqc committed:
Fixed dropping size 42 packets & IOError

Using IOError instead of catching socket.error as it's the parent
exception. Also, noticed that the data appeared valid for size 42 packets
coming from UART but were being dropped. This explained missing lines.
Adding logic to identify size 42 data strings and acting accordingly seems
to work and I can observe no missing data in my counting data tests
genering lists counting from 0 to 299.

kb1lqc commented:
OK @kb1lqd with https://github.com/FaradayRF/Faraday-Software/pull/251/commits/c1d692fa7cf3bd55dc4c1717ce6a3ae1afac4dba I appear to be getting all data sent over the original link. Below is a valid count between 0 and 299 in my test string of a list counting all 300 numbers:

![image](https://user-images.githubusercontent.com/331540/28859333-913ea560-770b-11e7-8ac1-209fd07c5621.png)

Not sure if this is what you intended but it appears to work. In the end I think the uart code needs to be shored up but this seems to work for now.

kb1lqc committed:
Removed poor taste comment/try

Yeah I'm a 12 year old sometimes :P

kb1lqc committed:
Added better inline comment and TODO

MSP430 header comment improved to note the header instead of preamble.
Also added a TODO reminding us of better handling the code. Not sure a
global variable is best here so will sit on it. Low risk.

kb1lqc committed:
Fixed connection bug in senddata

I connected in the while loop when I should not have. This was a bug
waiting to happen and also possibly part of the manifestation of requiring
the send loop to have a delay...

kb1lqc committed:
Added test code attempting to conn/disconn RX

I can't connect and disconnect a RX socket connection when no data was
ever received (so the dictionary item doesn't exist). This is a bug...

kb1lqc committed:
Pytest updates

kb1lqd commented:
@kb1lqc in regards to https://github.com/FaradayRF/Faraday-Software/pull/251#issuecomment-319569549

Yes, this is only done on the MSP430 firmware where the link between UART -> RF and RF-> UART occurs. The 128 byte packet is split into 3 RF packets and reassembled on the other side. The result is through UART the same item appears out of the RX Faraday unit. You should not see this in the software... Maybe something is getting through the logic...

kb1lqd commented:
@kb1lqc Seems too work better and not error out but I'm seeing issues with closing the RX client and then  trying too reconnect... The RX program just stays blank... No apparent error though.

![image](https://user-images.githubusercontent.com/339144/28864363-2aaf860a-7721-11e7-9355-23449a719ab2.png)

kb1lqc commented:
@kb1lqd it's possible I need to be tricky with my while loops. I apparently achieve it in [this commit earlier](https://github.com/kb1lqc/Faraday-Software/blob/77b57acd5aa608d60a76599f12171c14ded22ed8/faraday/proxy.py#L521). However, the problem seems to be arising from the fact that I need getDicts[unit][1] to exist but until data is RX'd it doesn't exist. When I apparently try to mitigate that I also am missing something that exits the thread so reconnection isn't possible. This is tricky...

kb1lqd commented:
OK @kb1lqc is this something you think you can achieve again? I think it is pretty critical considering every reconnect would need a restart of proxy. We're very close to closing this initial PR otherwise.

reillyeon commented:
It would be nice if this used a WebSocket instead of a regular TCP/IP socket so that it integrated into the existing REST API. Not something to be fixed in this PR but something to think about when we switch to a framework that can support WebSockets.

kb1lqc commented:
@reillyeon We locked at websockets but it appeared they were not backwards compatible with normal socket use.

kb1lqd commented:
@kb1lqc Do you think you can try to get the RX no-reconnect socket fixed? I may need to pre-approve before this weekend. Not a huge deal but certainly is an annoyance.

kb1lqc commented:
@kb1lqd Yeah that's my main goal right now. Pesky. Will continue it in the morning.

kb1lqc commented:
Quick look at it this morning and I notice the following:
* In my RX worker thread for clients receiving data over socket from RF reception I am having trouble properly closing sockets & waiting for next connection
* Connection seems to stay open and make the thread block

I'm struggling for time lately so hopefully I can get this done over the weekend. Did you want to approve under certain conditions be met? @kb1lqd? I can also just push it when ready with admin privileges though I am not the biggest fan of those.

kb1lqc committed:
Simplified RX worker code handling reconnects

Realized how I need to place the acceptConnection() function in the main
while loop. This allows the RX thread to wait for the next connection.
Just sends "Test..." repeatedly to connected thread.

kb1lqc committed:
Reconnects but exits when no data left in queue

Progress, need to figure out how to keep connection open until client
closes socket.

kb1lqc committed:
Reconnects seems to work w/ cont. data send

Manipulated the breaks and sending loops to get a continuous stream of
data and also multiple RX client reconnects. Great step forward. Now to
cleanup and optimize.

kb1lqc committed:
Cleaned up socket_Worker_RX

Removed old code and test stuff. Tested works as good as it did in prior
commit.

kb1lqc committed:
Missed commented out code in Socket_Worker_RX

Simple delete of commented out code

kb1lqc committed:
Cleaned up commented out code in sendData

Removed old commented out code and tested to see that socket functionality
works as it did prior.

kb1lqc committed:
Partial completed cleaning up SendData

Commenting and cleaning up code, touched code prior to unpacking data.

kb1lqc committed:
Cleaned up rest of sendData()

Basic cleanup complete. Seems to allow reconnection and sending of data on
both TX and RX clients. Good enough for now!

kb1lqc committed:
Removed test code for TX socket

Removed some leftover logger statements to clean up output.

kb1lqc committed:
Added helpful connect loggers for TX/RX ports

Displays information about connected sockets both TX and RX threads.

kb1lqc committed:
Closed RX socket when appropriate

Added a closeConnection() call to the end of the loop for
socket_Worker_RX()

kb1lqc committed:
Reworded connection information

Helpful connection information reworded to help clarify which threads were
connected.

kb1lqc committed:
Fixed pytest failures

kb1lqc commented:
Alright, as of https://github.com/FaradayRF/Faraday-Software/pull/251/commits/d5e5c30604ae5c29ea37216463d7669592e964aa I believe I fixed the client-side reconnect issues as well as cleaned up the code. Is it stellar code? No, but it's good enough right now.

Will perform some manual packaging tests and such before pulling this in per @kb1lqd verbal approval already obtained (he's on vacation) contingent on the client reconnection issue being mitigated.

@reillyeon @el-iso

kb1lqd commented:
Awesome! I have no computer do you're on your own. Good luck, we're all
counting on you.

On Aug 6, 2017 8:55 PM, "Bryce Salmi" <notifications@github.com> wrote:

Alright, as of d5e5c30
<https://github.com/FaradayRF/Faraday-Software/commit/d5e5c30604ae5c29ea37216463d7669592e964aa>
I believe I fixed the client-side reconnect issues as well as cleaned up
the code. Is it stellar code? No, but it's good enough right now.

Will perform some manual packaging tests and such before pulling this in
per @kb1lqd <https://github.com/kb1lqd> verbal approval already obtained
(he's on vacation) contingent on the client reconnection issue being
mitigated.

@reillyeon <https://github.com/reillyeon> @el-iso
<https://github.com/el-iso>

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/FaradayRF/Faraday-Software/pull/251#issuecomment-320562973>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/AAUsyMEPxoXSx0kfE69bgN7kxUbIG3b1ks5sVoqlgaJpZM4OpM7b>
.

kb1lqc committed:
Updated to rev 0.0.1016 for Data Socket

Rev 0.0.1016 will include data socket functionality. Still very much
development.

kb1lqc commented:
Tested by hand and appears to work as expected:
* Client Transmission program can connect to a socket and send arbitrary data
* Client Reception program can connect to a socket and receive arbitrary data
* Works correctly with two Faraday radios connected via USB to a single Proxy
* Encodes IP ports used for sockets in `proxy.ini`
 * Hard-codes RX client socket as Faraday radio TX client socket + 10
* Reconnections with clients possible
* Connected RX client with no data ever received on RX Faraday radio works fine

Please note:
* This PR pulls in extremely development level functionality. We will use subsequent PR's to improve it
* Documentation is sparse at this time
* All Faraday's broadcast data per firmware so any Faraday within range WILL receive any data sent
* You must provide your own fault detection and correction code at a higher level when communicating with the socket.

Will use my admin privileges to merge.

Thanks @kb1lqd!

@reillyeon @el-iso
